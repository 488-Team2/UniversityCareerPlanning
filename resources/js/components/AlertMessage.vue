<template>
    <div>
        <div :class="alertType" class='alert' role="alert" v-for="alert in alertArray">
            {{ alert }}
        </div>
    </div>
</template>

<script>
export default {
    name: "alert-message",
    props: {
        alertArray: Array,
        alertType: String
    }
}
</script>

<style scoped>

</style>
