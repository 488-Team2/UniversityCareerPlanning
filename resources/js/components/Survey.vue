<template>
    <div>
        <alert :alert-array="alerts" :alertType="alertType"></alert>
        <h2 :class="currentQuestion.question_code">{{ currentQuestion.question_text }}</h2>
        <div>
            <div>
                <label>
                    <input v-model="selectedAnswer" name="answer" type="radio"
                           value="yes">
                    Yes
                </label>
            </div>
            <div>
                <label>
                    <input v-model="selectedAnswer" name="answer" type="radio"
                           value="no">
                    No
                </label>
            </div>
        </div>

        <p>Current question: {{ this.currentQuestionIndex + 1 }}</p>
        <button v-on:click="goToNextQuestion" :disabled="isDisabled">{{ buttonText }}</button>
    </div>
</template>

<script>
export default {
    name: "Survey",
    data() {
        return {
            questions: Array,
            currentQuestion: {},
            currentQuestionIndex: 0,
            selectedAnswer: null,
            responses: [],
            alerts: [],
            alertType: "",
            responseArray: {"R": 0, "I": 0, "A": 0, "S": 0, "E": 0, "C": 0}
        }
    },
    created() {
        this.fetchQuestions();
    },
    methods: {
        fetchQuestions() {
            let self = this;
            axios({
                method: "get",
                Accept: "application/json",
                url: 'api/survey/questions',
            }).then(function (response) {
                self.questions = response.data.data;
                self.currentQuestion = self.questions[0];
                console.log(self.currentQuestion);
            }).catch(error => console.log(error));
        },
        goToNextQuestion() {
            if (this.questions.length > 0 && this.currentQuestionIndex !== this.questions.length - 1) {
                if (this.selectedAnswer === "yes")
                    this.responseArray[this.currentQuestion.question_code]++;

                this.responses.push(this.selectedAnswer);
                this.currentQuestion = this.questions[++this.currentQuestionIndex];
                this.selectedAnswer = null;
            } else {
                //Grade the responses and give a recommendation for the types of degrees
                this.alerts = [];
                this.alerts.push('Successfully submitted career survey!');
                this.alertType = "alert-success";
            }
        }
    },
    computed: {
        isDisabled() {
            return this.selectedAnswer === null;
        },
        buttonText() {
            if (this.currentQuestionIndex < this.questions.length - 1) {
                return "Next question";
            } else {
                return "Complete survey";
            }
        }
    }
}
</script>

<style scoped>

</style>
