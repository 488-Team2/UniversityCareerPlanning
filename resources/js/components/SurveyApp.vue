<template>
    <div>
        <component v-bind:is="currentComponent" @submitSurvey="changeSurveyView" @beginSurvey="beginSurveyQuiz" v-bind="currentProperties"></component>
    </div>
</template>

<script>
export default {
    name: "SurveyApp",
    data() {
        return {
            currentComponent: 'survey-start',
            responses: []
        }
    },
    methods: {
        changeSurveyView(responses) {
            this.currentComponent = 'survey-results';
            this.responses = responses;
            confetti();
        },
        beginSurveyQuiz() {
            this.currentComponent = 'survey';
        }
    },
    computed: {
        currentProperties() {
            if (this.currentComponent === 'survey-results') {
                return {responses: this.responses}
            }
        }
    }
}
</script>

<style scoped>

</style>
