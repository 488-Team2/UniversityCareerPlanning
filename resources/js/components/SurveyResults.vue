<template>
    <div>
        <h2>Your career survey results</h2>
        <ul>
            <li v-for="degree in this.degreeOptions">
                {{ degree.degree_name }} - {{ degree.degree_description }}
            </li>
        </ul>
    </div>
</template>

<script>
export default {
    name: "surveyResults",
    props: {
        responses: Object
    },
    data() {
        return {
            topLetters: Object,
            degreeOptions: Array
        }
    },
    created() {
        this.gradeSurvey();
    },
    methods: {
        gradeSurvey() {
            this.topLetters = Object.fromEntries(Object.entries(this.responses).sort(([, a], [, b]) => b - a).slice(0, 3));
            console.log(this.topLetters);
            this.showDegreeOptions();
        },
        showDegreeOptions() {
            let self = this;
            let degreeUrl = "/api/degrees/codes/";
            for (let topLettersKey in this.topLetters) {
                degreeUrl += topLettersKey + " ";
            }

            axios.get(degreeUrl)
                .then(function (response) {
                    self.degreeOptions = response.data.data;
                });
        }
    }
}
</script>

<style scoped>

</style>
