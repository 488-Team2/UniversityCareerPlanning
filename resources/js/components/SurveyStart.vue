<template>
    <div>
        <h1>Career Survey</h1>
        <p>Welcome to the career survey. This survey will ask a series of questions, and your responses to those questions
            will help us to recommend some potential degree options. When you have finished answering the questions,
            click on 'complete survey' and the quiz will grade your
            responses, and recommend degree offerings that may be a good fit for you.</p>
        <div class="text-center">
            <button class="btn btn-primary" v-on:click="beginSurvey">Begin career survey</button>
        </div>
    </div>
</template>

<script>
export default {
    name: "SurveyStart",
    methods: {
        beginSurvey() {
            this.$emit('beginSurvey');
        }
    }
}
</script>

<style scoped>
.h1{
    font-size:60px;
    text-align:center;
}
</style>
