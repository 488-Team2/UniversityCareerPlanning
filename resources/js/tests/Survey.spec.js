import {mount} from '@vue/test-utils'
import SurveyApp from "../components/SurveyApp.vue";
import Survey from "../components/Survey.vue";
import axios from "axios";
import SurveyResults from "../components/SurveyResults.vue";
import SurveyStart from "../components/SurveyStart.vue";
import AlertMessage from "../components/AlertMessage.vue";

test('has question text', async () => {
    const surveyWrapper = mount(Survey, {components: {AlertMessage}});
    const response = await axios.get('/api/survey/questions')
    expect(surveyWrapper.vm.$data.questions[0].question_text).toEqual(response.data.data[0].question_text);
});

test('Should load question data into questions object', async () => {
    const surveyWrapper = mount(Survey, {components: {AlertMessage}});
    await surveyWrapper.vm.fetchQuestions();
    expect(surveyWrapper.vm.$data.questions.length).toBeGreaterThan(0);
});

test('Advances survey when user clicks response and selects next question button', async () => {
    const surveyWrapper = mount(Survey, {components: {AlertMessage}});
    expect(surveyWrapper.exists()).toBe(true);
    await surveyWrapper.vm.fetchQuestions();
    const currentIndex = surveyWrapper.vm.$data.currentQuestionIndex;
    let answerInput = surveyWrapper.find('input[type=radio]');
    let nextQuestionButton = surveyWrapper.find('.btn');

    await answerInput.trigger('change');
    await nextQuestionButton.trigger('click');
    expect(surveyWrapper.vm.$data.currentQuestionIndex).toBeGreaterThan(currentIndex);
});

test('Button stays disabled when answer is not selected', async () => {
    const surveyWrapper = mount(Survey, {components: {AlertMessage}});
    await surveyWrapper.vm.fetchQuestions();
    const currentIndex = surveyWrapper.vm.$data.currentQuestionIndex;
    let nextQuestionButton = surveyWrapper.find('.btn');

    await nextQuestionButton.trigger('click');
    expect(surveyWrapper.vm.$data.currentQuestionIndex).toEqual(currentIndex);
    expect(surveyWrapper.vm.$data.currentQuestion.question_text).toEqual(surveyWrapper.vm.$data.questions[0].question_text);
});

test('Displays career survey results after answering all questions', async () => {
    const wrapper = mount(SurveyApp, {
        components: {
            SurveyStart, Survey, SurveyResults, AlertMessage: {
                name: "alert-message"
            }
        }
    });
    const surveyStartWrapper = wrapper.findComponent(SurveyStart);
    await surveyStartWrapper.find('button').trigger('click');
    const surveyWrapper = wrapper.findComponent(Survey);
    expect(surveyWrapper.exists()).toBe(true);
    await surveyWrapper.vm.fetchQuestions();
    const initialQuestionIndex = surveyWrapper.vm.$data.currentQuestionIndex;
    let answerInput = surveyWrapper.find('input[type=radio]');
    let nextQuestionButton = surveyWrapper.find('.btn');

    let questionsArrayLength = surveyWrapper.vm.$data.questions.length;

    for (let i = 0; i < questionsArrayLength; i++) {
        await answerInput.trigger('change');
        await nextQuestionButton.trigger('click');
    }

    expect(surveyWrapper.vm.$data.currentQuestionIndex).toEqual(questionsArrayLength - 1);
    expect(surveyWrapper.vm.$data.currentQuestionIndex).toBeGreaterThan(initialQuestionIndex);
    const resultsWrapper = wrapper.findComponent(SurveyResults);
});

test('Loads SurveyStart child component', async () => {
    const wrapper = mount(SurveyApp, {
        components: {
            SurveyStart, Survey, AlertMessage: {
                name: "alert-message"
            }
        }
    });
    const surveyStartWrapper = wrapper.findComponent(SurveyStart);
    expect(surveyStartWrapper.exists()).toBe(true);
});
