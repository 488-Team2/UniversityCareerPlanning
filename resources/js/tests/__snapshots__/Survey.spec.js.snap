// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uses mounts 1`] = `
VueWrapper {
  "__app": Object {
    "_component": Object {
      "__emits": Object {},
      "__props": Array [
        Object {},
        Array [],
      ],
      "name": "VTU_ROOT",
      "render": [Function],
    },
    "_container": <div
      data-v-app=""
    >
      <div>
        <alert
          alert-array=""
          alerttype=""
        />
        <h2
          class=""
        />
        <div
          class="answer-inputs"
        >
          <div>
            <label>
              <input
                name="answer"
                type="radio"
                value="yes"
              />
               Yes 
            </label>
          </div>
          <div>
            <label>
              <input
                name="answer"
                type="radio"
                value="no"
              />
               No 
            </label>
          </div>
        </div>
        <p>
          Current question: 1
        </p>
        <button
          class="btn btn-primary"
          disabled=""
        >
          Complete survey
        </button>
      </div>
    </div>,
    "_context": Object {
      "app": [Circular],
      "components": Object {
        "transition": Object {
          "name": "transition",
          "props": undefined,
          "render": [Function],
        },
        "transition-group": Object {
          "name": "transition-group",
          "props": undefined,
          "render": [Function],
        },
      },
      "config": Object {
        "errorHandler": undefined,
        "globalProperties": Object {},
        "isCustomElement": [Function],
        "isNativeTag": [Function],
        "optionMergeStrategies": Object {},
        "performance": false,
        "warnHandler": undefined,
      },
      "directives": Object {},
      "mixins": Array [
        Object {
          "__emits": null,
          "__props": Array [],
          "beforeCreate": [Function],
        },
      ],
      "provides": Object {},
      "reload": [Function],
    },
    "_props": null,
    "_uid": 0,
    "component": [Function],
    "config": Object {
      "errorHandler": undefined,
      "globalProperties": Object {},
      "isCustomElement": [Function],
      "isNativeTag": [Function],
      "optionMergeStrategies": Object {},
      "performance": false,
      "warnHandler": undefined,
    },
    "directive": [Function],
    "mixin": [Function],
    "mount": [Function],
    "provide": [Function],
    "unmount": [Function],
    "use": [Function],
    "version": "3.0.6",
  },
  "__setProps": [Function],
  "componentVM": Object {
    "alertType": "",
    "alerts": Array [],
    "buttonText": "Complete survey",
    "currentQuestion": Object {},
    "currentQuestionIndex": 0,
    "hasOwnProperty": [Function],
    "isDisabled": true,
    "questions": [Function],
    "responseArray": Object {
      "A": 0,
      "C": 0,
      "E": 0,
      "I": 0,
      "R": 0,
      "S": 0,
    },
    "responses": Array [],
    "selectedAnswer": null,
  },
  "rootVM": Object {},
  "wrapperElement": <div>
    <alert
      alert-array=""
      alerttype=""
    />
    <h2
      class=""
    />
    <div
      class="answer-inputs"
    >
      <div>
        <label>
          <input
            name="answer"
            type="radio"
            value="yes"
          />
           Yes 
        </label>
      </div>
      <div>
        <label>
          <input
            name="answer"
            type="radio"
            value="no"
          />
           No 
        </label>
      </div>
    </div>
    <p>
      Current question: 1
    </p>
    <button
      class="btn btn-primary"
      disabled=""
    >
      Complete survey
    </button>
  </div>,
}
`;
